### ISSUE TITLE
Introducing Refactoring

### ISSUE THREAD

**Issue by davidhalter** [davidhalter/jedi#103]
I'm thinking about introducing basic refactoring methods for Jedi. Apart from renaming (which is already possible, there's not really much I know about typical/usual refactoring methods.

The goal would be to add a new module `refactoring` that stands for itself and returns the changed lines of different files (which is like a unified interface for all refactoring operations). Maybe it should also return changed file names.

So, @dbrgn, @tkf, @jjay or anyone else:
1. Can you tell me what you think are the most basic refactoring operations (I'm speaking of Python, of course. Clearly some refactoring methods in Java are possible but maybe make no sense for Python).
2. Does anyone want to get involved in this? As far as I think it will be really different from the Jedi core and mostly self-explanatory.

**Comment by davidhalter** [comment]
Everything we have done in the refactoring module can be rewritten. With the new parser it should be way easier. Just start anew.

**Comment by asmeurer** [comment]
A little tutorial on how to use the parser would be helpful, or maybe a little toy refactoring example. I've been playing around with it, but mostly by examining the objects interactively. But I'm not sure if I'm using any non-public interfaces or doing things the wrong way. 

For instance, what is the correct way to combine multiple ASTs? I've been mutating the `children` attribute, but I'm not sure if that's correct. For instance, you can easily create ASTs that round-trip in terms of AST -> get_code -> AST because the original AST is not achievable in the grammar, like

```py
In [2]: from jedi.parser import *

In [3]: grammar = load_grammar()

In [15]: p = Parser(grammar, "(x)")

In [20]: i = Parser(grammar, 'f')

In [22]: i.module.children
Out[22]: [Node(simple_stmt, [<Name: f@1,0>, <Whitespace: >]), <Whitespace: >]

In [24]: p.module.children
Out[24]:
[Node(simple_stmt, [Node(atom, [<Operator: (>, <Name: x@1,1>, <Operator: )>]), <Whitespace: >]),
 <Whitespace: >]

In [26]: p.module.children.insert(0, i.module.children[0])

In [27]: p.module.get_code()
Out[27]: 'f(x)'

In [28]: p.module.children
Out[28]:
[Node(simple_stmt, [<Name: f@1,0>, <Whitespace: >]),
 Node(simple_stmt, [Node(atom, [<Operator: (>, <Name: x@1,1>, <Operator: )>]), <Whitespace: >]),
 <Whitespace: >]

In [29]: r = Parser(grammar, 'f(x)')

In [30]: r.module.children
Out[30]:
[Node(simple_stmt, [Node(power, [<Name: f@1,0>, Node(trailer, [<Operator: (>, <Name: x@1,2>, <Operator: )>])]), <Whitespace: >]),
 <Whitespace: >]
```

Am I doing things how you would do them here? The basic question I want to answer here is, how can I build up and manipulate code entirely from the AST?

**Comment by asmeurer** [comment]
Aside from the fact that my `In [28]` uses a `Name` node next to an `atom` node instead of a `power` node (whatever that means), there is the issue of the column number on `x` not being correct any more. Is that something I need to worry about? `get_code` seems to have done the right thing, which I'm impressed by, but I haven't played with it enough to know if things will always work.

**Comment by davidhalter** [comment]
TL;DR *Positions don't matter, use `Node.children`, `Leaf.value` and `Leaf.prefix` to modify an AST.

> But I'm not sure if I'm using any non-public interfaces or doing things the wrong way. 

You did just fine :) As you can see, there's mostly `Node.children` that you can play around with. 

> For instance, what is the correct way to combine multiple ASTs?

I've never thought about this. But it doesn't really matter, the AST can and should be changed however you want. In the end you want a good `get_code` output, as long as that's correct you're fine.

> Am I doing things how you would do them here? The basic question I want to answer here is, how can I build up and manipulate code entirely from the AST? 

You're doing it exactly the way I would do it. The only thing that is not going to work with this is that the start/end positions are going to be wrong (`start_pos`, `end_pos`). I wouldn't care about this, since after combining, you usually don't need these positions anymore. `get_code` doesn't use those positions at all. It just uses `Leaf.value` and `Leaf.prefix`. That's all.

For example if I wanted to replace a `Name` node with a certain piece of code, I wouldn't even care about the type, just say `name.value = '()'` and call `get_code` after that:

```
>>> from jedi.parser import Parser, load_grammar
>>> p = Parser(load_grammar(), 'foo = 3')
>>> p.module.children[0].children[0]
<ExprStmt: foo = 3@1,0>
>>> name = p.module.children[0].children[0].children[0]
>>> name
<Name: foo@1,0>
>>> name.value = 'bar[0]'  # name is really not a name anymore, but we don't care.
>>> p.module.get_code()
'bar[0] = 3'
```

**Comment by davidhalter** [comment]
Closed in favor of #667 that summarizes and gives ideas on how to do it.
